### PR TITLE
Port keybind improvements from tg, use them to fix arm augment issues

### LINKED ISSUE #5618: Tool arm implant welder falling off
## Reproduction:
Honestly forgot about this until I found that I took a [screenshot of it](https://imgur.com/gixRT4P).
I just used the welder attachment as normal, but wasn't able to retract it back into the implant. I then called out another attachment and the welder was dropped from my hand. I was able to use the welder still, but I wasn't able to recall it and was actually able to even put it in my backpack.

### PR DESCRIPTION
**Pull request by KubeRoot**
## About The Pull Request

Ports the following PRs:
- https://github.com/tgstation/tgstation/pull/49116
- https://github.com/tgstation/tgstation/pull/52219
- https://github.com/tgstation/tgstation/pull/54914

Fixes https://github.com/BeeStation/BeeStation-Hornet/issues/5618

**I only did rudimentary testing on this.** As it affects a basic element of the game (keybindings) I would recommend a testmerge.

The first PR adds a `can_use` proc to `keybinding` datums, which is called to check if a keybinding is valid to be used before it's executed. Checks shared between keybindings are moved to this new proc, defined on a base type, to reduce boilerplate and help avoid issues due to people forgetting to include those checks.

The second PR adds signals (`keybind_signal`) for every `keybinding`. The signals are called in the base `/datum/keybinding/down()`, which needs to call parent, and can cancel execution of overriden `down` definitions. The signals are required to be defined on every keybinding.
Also included is a unit test that checks to ensure every keybinding with a `name` defined also has a `keybind_signal` defined.
**NOTE:** This only applies to serverside keybindings. This does not include, for example, `T` for say, `Tab` for hotkey mode, `O` for OOC.

The third PR fixes issues with being able to detach arm augment items from your hand. When I was porting the arm augment rework, keybind signals were a missing dependency, and I went with a suggestion I saw on in a comment on that PR. As it turns out, this caused issues that went unnoticed for a while, allowing you to easily detach your augment items from your hand.  

## Why It's Good For The Game

Bugfixes are good, keybind signals might be useful in other places.

## Changelog
:cl: KubeRoot, AnturK, Coul, Rohesie, zxaber
fix: Fixed at least two ways of detaching arm augment items from your hand.
code: Added can_use and signals to keybinds.
/:cl:

### CODE REVIEW

**Comment by Howluinb** [comment]
since you're working so hard on keybind, would it be possible for you to tackle the *very* long standing issue that is the inability to bind movement keys, to any other keys?

It'd be extremely loved from any person in the community with a non qwerty keyboard

https://github.com/BeeStation/BeeStation-Hornet/issues/3633

**Comment by KubeRoot** [comment]
> since you're working so hard on keybind, would it be possible for you to tackle the _very_ long standing issue that is the inability to bind movement keys, to any other keys?

I might look into that, I think tg has support for macro-based rebindable keybinds, but I think it might be best to leave it for a separate PR.

This PR currently only does code-side stuff and fixes - refactoring keybinds slightly, adding the signals, and using them to fix a bug.
Adding rebindable movement keys would be a new feature that would require reworking how those keybinds work, add them to UI, and could possibly break more badly.

File: code/datums/keybinding/mob.dm [hunk 1]
```diff
@@ -154,9 +198,13 @@
 	name = "toggle_move_intent_alt"
 	full_name = "press to cycle move intent"
 	description = "Pressing this cycle to the opposite move intent, does not cycle back"
+	keybind_signal = COMSIG_KB_MOB_TOGGLEMOVEINTENTALT_DOWN
 
 /datum/keybinding/mob/toggle_move_intent_alternative/down(client/user)
 	. = ..()
+	if(.)
+		return
+	. = ..()
```
Comment (francinum) [inline code/datums/keybinding/mob.dm]: what the actual fuck

Comment (KubeRoot) [reply code/datums/keybinding/mob.dm]: Oh shit

Edit: Oh man, looks like whoever coded this didn't realise we didn't have a prerequisite. I believe the code wasn't doing any checks for if you should actually be able to use the keybind. I'm gonna webedit this for now, will test tomorrow if I remember.

**Review by Archanial — APPROVED** [review]

**Pull request merged**